liberum_vault/src/vault.rs

```diff
@@ -0,0 +1,481 @@
+    vault_dir_path: PathBuf,
+}
+
+type FragmentData = BoxStream<'static, Result<Bytes, io::Error>>;
+
+pub struct StoreFragment(Option<Key>, FragmentData);
+pub struct LoadFragment(Key);
+
+impl Vault {
+    const DEFAULT_VAULT_DATABASE_NAME: &'static str = "vault.db3";
+    const FRAGMENT_DIR_NAME: &'static str = "fragments";
+    const TEMP_DIR_NAME: &'static str = "temp";
+    const MIN_FRAGMENT_SIZE: u64 = 4096;
+
+    pub async fn new(vault_dir_path: &Path) -> Result<Vault> {
+        Self::ensure_dirs(vault_dir_path).await?;
+
+        let db_path = Self::default_db_path(vault_dir_path);
+        let db = Connection::open(db_path).await?;
+
+        Ok(Vault {
+            db,
+            vault_dir_path: vault_dir_path.to_path_buf(),
+        })
+    }
+
+    pub async fn fragment(path: &Path) -> Result<Vec<FragmentData>> {
+        let file_size = tokio::fs::metadata(path).await?.len();
+        let fragment_sizes = Self::fragment_sizes(file_size);
+        let mut current_pos = 0;
+        let mut result: Vec<FragmentData> = Vec::new();
+
+        for fragment_size in fragment_sizes {
+            let mut f = File::open(path).await?;
+            f.seek(std::io::SeekFrom::Start(current_pos)).await?;
+
+            let buf_reader = BufReader::new(f);
+            let reader_stream = ReaderStream::new(buf_reader.take(fragment_size));
+
+            result.push(reader_stream.boxed());
+            current_pos += fragment_size;
+        }
+
+        Ok(result)
+    }
+
+    fn fragment_sizes(target: u64) -> Vec<u64> {
+        let mut fragment_sizes = Vec::new();
+        let mut current_size = cmp::max(Self::MIN_FRAGMENT_SIZE, Self::power_2_upto(target));
+        let mut current_target = target;
+
+        while current_target != 0 {
+            current_size = cmp::max(
+                Self::MIN_FRAGMENT_SIZE,
+                Self::power_2_desc_from_power_2(current_size, current_target),
+            );
+
+            fragment_sizes.push(current_size);
+            current_target = current_target.saturating_sub(current_size);
+        }
+
+        fragment_sizes
+    }
+
+    fn power_2_upto(limit: u64) -> u64 {
+        let powers_from_1 = successors(Some(1u64), |&n| Some(n * 2))
+            .take_while(|x| x <= &limit)
+            .last();
+
+        once(0u64).chain(powers_from_1).last().unwrap()
+    }
+
+    fn power_2_desc_from_power_2(start: u64, limit: u64) -> u64 {
+        assert!(Self::is_power_of_2(start));
+
+        if limit == 0 {
+            return 0;
+        }
+
+        successors(Some(start), |&n| Some(n / 2))
+            .take_while(|x| x > &(limit / 2) || x == &limit)
+            .last()
+            .unwrap()
+    }
+
+    fn is_power_of_2(number: u64) -> bool {
+        number > 0 && (((number) & (number - 1)) == 0)
+    }
+
+    async fn prepare_db(&self) -> Result<()> {
+        const CREATE_FRAGMENT_TABLE_QUERY: &str = "
+            CREATE TABLE IF NOT EXISTS fragment (
+                hash0 INTEGER,
```

Comment:
I think NOT NULL has to be set explicitly in sqlite
`hash0 INTEGER,` > `hash0 INTEGER NOT NULL,`

Reply:
Fixed.